[quant][core][gpu][improvement] Modified quantized cudnn linear caching

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75447

Summary plan:
The previous CacheKey definition for cudnn quantized linear operator was insufficient.
This PR expands upon the definition by adding additional necessary parameters in CacheKey,
(e.g., input, weight size, kReluFused)

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D35473728](https://our.internmc.facebook.com/intern/diff/D35473728)

dzdang committed:
[quant][core][gpu][improvement] Modified quantized cudnn linear caching

Summary plan:
The previous CacheKey definition for cudnn quantized linear operator was insufficient.
This PR expands upon the definition by adding additional necessary components,
e.g., input and weight size

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][improvement] Modified quantized cudnn linear caching"


Summary plan:
The previous CacheKey definition for cudnn quantized linear operator was insufficient.
This PR expands upon the definition by adding additional necessary components,
e.g., input and weight size

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D35473728](https://our.internmc.facebook.com/intern/diff/D35473728)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35473728).

dzdang committed:
Update on "[quant][core][gpu][improvement] Modified quantized cudnn linear caching"


Summary plan:
The previous CacheKey definition for cudnn quantized linear operator was insufficient.
This PR expands upon the definition by adding additional necessary components,
e.g., input and weight size

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D35473728](https://our.internmc.facebook.com/intern/diff/D35473728)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][improvement] Modified quantized cudnn linear caching"


Summary plan:
The previous CacheKey definition for cudnn quantized linear operator was insufficient.
This PR expands upon the definition by adding additional necessary components,
e.g., input and weight size

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D35473728](https://our.internmc.facebook.com/intern/diff/D35473728)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35473728).

dzdang committed:
Update on "[quant][core][gpu][improvement] Modified quantized cudnn linear caching"


Summary plan:
The previous CacheKey definition for cudnn quantized linear operator was insufficient.
This PR expands upon the definition by adding additional necessary components,
e.g., input and weight size

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D35473728](https://our.internmc.facebook.com/intern/diff/D35473728)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35473728).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -31,16 +32,49 @@ cudnn_frontend::MatMulDesc_v8 getLinearDescriptor(cudnnDataType_t dataType) {
+// this can be increased, if necessary
+constexpr uint8_t max_num_input_dim = 5;
+struct LinearParams {
+  c10::DeviceIndex device_id;
```
can you explain why do we need `device_id`?

dzdang replied:
I'm not sure, but it was in conv too so I ported it over. see:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/ConvShared.h#L24
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/ConvShared.cpp#L90

maybe the workflow could transfer devices and we want to make sure we only use the cached result for the same device?

jerryzh168 replied:
I see, sounds good

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -31,16 +32,49 @@ cudnn_frontend::MatMulDesc_v8 getLinearDescriptor(cudnnDataType_t dataType) {
+constexpr uint8_t max_num_input_dim = 5;
+struct LinearParams {
+  c10::DeviceIndex device_id;
+  cudnnDataType_t dataType;
```
and does this ever change?

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -273,6 +308,9 @@ at::Tensor PackedLinearWeightCudnn::apply_impl(
     double output_scale,
     int64_t output_zero_point) {
+  TORCH_CHECK(act.dim() <= max_num_input_dim, "Number of dimensions for input tensor exeeds ", max_num_input_dim,
+  "; Increase max_num_input_dim in aten/src/ATen/native/quantized/cudnn/Linear.cpp to a larger value, if necessary, and recompile.");
```
probably shouldn't say this, since TORCH_CHECK errors are the user errors, it happens when user didn't provide proper arguments, I think we can just remove this line. we should be able to figure out the constraint and increase this number

dzdang replied:
removed

dzdang committed:
Update on "[quant][core][gpu][improvement] Modified quantized cudnn linear caching"


Summary plan:
The previous CacheKey definition for cudnn quantized linear operator was insufficient.
This PR expands upon the definition by adding additional necessary components,
e.g., input and weight size

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D35473728](https://our.internmc.facebook.com/intern/diff/D35473728)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35473728).

dzdang committed:
Update on "[quant][core][gpu][improvement] Modified quantized cudnn linear caching"


Summary plan:
The previous CacheKey definition for cudnn quantized linear operator was insufficient.
This PR expands upon the definition by adding additional necessary parameters in CacheKey,
(e.g., input, weight size, kReluFused)

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D35473728](https://our.internmc.facebook.com/intern/diff/D35473728)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35473728).

dzdang commented:
@jerryzh168  also this one

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -31,16 +32,50 @@ cudnn_frontend::MatMulDesc_v8 getLinearDescriptor(cudnnDataType_t dataType) {
+void setLinearParams(
+    LinearParams* params, const at::Tensor& input, const at::Tensor& weight,
+    bool deterministic, bool allow_tf32) {
+  // operator datatype needs to be int32 for int8 matmul, but we can
+  // set the datatype for output tensor to int32 or fp32
```
how is this comment related to the code?

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -31,16 +32,50 @@ cudnn_frontend::MatMulDesc_v8 getLinearDescriptor(cudnnDataType_t dataType) {
+  memset(params, 0, sizeof(LinearParams));
+  params->device_id = at::cuda::current_device();
+  params->dataType = CUDNN_DATA_INT32;
+  // ASSERT(weight.dim() == input.dim())
```
remove?

dzdang committed:
Update on "[quant][core][gpu][improvement] Modified quantized cudnn linear caching"


Summary plan:
The previous CacheKey definition for cudnn quantized linear operator was insufficient.
This PR expands upon the definition by adding additional necessary parameters in CacheKey,
(e.g., input, weight size, kReluFused)

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D35473728](https://our.internmc.facebook.com/intern/diff/D35473728)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35473728).

dzdang committed:
Update on "[quant][core][gpu][improvement] Modified quantized cudnn linear caching"


Summary plan:
The previous CacheKey definition for cudnn quantized linear operator was insufficient.
This PR expands upon the definition by adding additional necessary parameters in CacheKey,
(e.g., input, weight size, kReluFused)

Test plan:
```
python test/test_quantization.py TestQuantizedLinear.test_qlinear_cudnn
```

Differential Revision: [D35473728](https://our.internmc.facebook.com/intern/diff/D35473728)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35473728).
